stratum/hal/lib/barefoot/bf_pal_wrapper.cc

```diff
@@ -280,6 +282,46 @@ ::util::Status BFPalWrapper::PortLoopbackModeSet(int unit, uint32 port_id,
   return ::util::OkStatus();
 }
 
+namespace {
+
+// PortKey uses three possibile values for channel:
+//   > 0: port is channelized (first channel is 1)
+//   0: port is not channelized
+//   < 0: port channel is not important (e.g. for port groups)
+int GetZeroBasedChannel(const PortKey* port_key) {
+  if (port_key->channel > 0) {
+    // Convert base-1 channel to base-0 channel
+    return port_key->channel - 1;
+  }
+  // Non-channelized port uses channel 0, or return non-zero
+  return port_key->channel;
+}
+
+}  // namespace
+
+::util::Status BFPalWrapper::PortIdFromPortKeyGet(int unit, PortKey port_key,
+                                                  uint32* sdk_port_id) {
+  int channel = GetZeroBasedChannel(&port_key);
+  if (channel < 0) {
+    RETURN_ERROR(ERR_INVALID_PARAM) << "Channel must be set for port "
+        << port_key.port << " on dev " << unit << ".";
+  }
+  std::string port_str = absl::StrFormat("%d/%d", port_key.port, channel);
+  bf_dev_port_t dev_port;
+  auto bf_status = bf_pal_port_str_to_dev_port_map(
+      static_cast<bf_dev_id_t>(unit),
+      const_cast<char*>(port_str.c_str()),
```

Comment:
Don't.
https://stackoverflow.com/questions/56434272/why-is-the-const-overload-of-stdstringdata-still-restricted-in-modern-c/56434399#56434399

Reply:
I was wondering if you'd catch this. :)

Reply:
There is `string_as_array` or `&str[0]`.

Reply:
replaced with `snprintf`